runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/RegisterNodeFunction.java

```diff
@@ -340,6 +335,53 @@ public Node apply(MutableNetwork<Node, Edge> input) {
             }
             ptransformIdToPCollectionViews.put(ptransformId, pcollectionViews.build());
 
+            // This gets the main input pcollection id for this PTransform. This will use the id to
+            // retrieve the Key coder to give to the timer.
+            String mainInputKeyCoderId = "";
+            for (Node predecessorOutput : input.predecessors(node)) {
+              String mainInputPCollectionId = nodesToPCollections.get(predecessorOutput);
```

Comment:
Obtaining mainInputPCollectionId is the only step you need to do inside loop.
Extracting mainInputKeyCoderId given mainInputPCollectionId can be done outside of loop.